bluefin-internal/src/Bluefin/Internal/Handle.hs

```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE
+  KindSignatures,
+  RankNTypes,
+  TypeOperators #-}
+
+module Bluefin.Internal.Handle
+  ( -- * Effect signatures
+    Sig
+  , CovariantSig(..)
+
+    -- * Handle usage
+  , Handle
+
+    -- * Handle creation
+  , HandleImpl
+  , with
+  ) where
+
+import Data.Kind (Type)
+import Bluefin.Internal (Eff, mergeEff, useImpl, type (:&), type (:>))
+
+-- | An effect signature declares a set of effectful operations.
+--
+-- Concretely, it should be a record of functions parameterized by a monad
+-- which occurs in the codomain of every field.
+--
+-- A handle of signature @f@ can be defined simply as a value of type @f (Eff s)@.
+-- But to use it, a handle will have type @'Handle' f@ so that its operations
+-- can be invoked with different effect rows in 'Eff', as long as they contain
+-- the row @s@ in the type of its original definition @f (Eff s)@.
+-- See 'Handle' and 'with' for more detaills.
+--
+-- === Example
+--
+-- Effect with the two state operations @get@ and @put@:
+--
+-- @
+-- type State :: Type -> 'Sig'
+-- data State s m = MkState
+--   { get :: m s
+--   , put :: s -> m ()
+--   }
+-- @
+type Sig = (Type -> Type) -> Type
+
+-- | Effect signatures must be instances of this class: they are higher-order functors.
+class CovariantSig (f :: Sig) where
+  smap :: (forall x. m x -> n x) -> f m -> f n
```

Comment:
This is interesting. Compare:

```.hs
smap useImpl :: (CovariantSig f, e :> es) => f (Eff e) -> f (Eff es)

mapHandle :: (e :> es) => h e -> h es
```

So they basically do the same job. `class Handle`/`mapHandle` has the benefit that the `Eff` is rolled in to the definition, with the hope that there's less that can "go wrong" when trying to implement or use instances. `CovariantSig`/`smap` has the benefit of being "more standard", because it doesn't mention anything Bluefin-specific. I'm not sure what's preferable. The latter allows `f` to be reused across different effect library implementations.

Reply:
I would be in favor of reusing what's already here. There are two things called `Handle` now. Should we rename one?

Reply:
Yeah, I don't know which one though. Your one sounds like it deserves the name more. Perhaps the class should be called `IsHandle`.

Reply:
Then we'd have something like `instance IsHandle (Handle f)` which seems pretty reasonable!

Reply:
I just observed that there's stuff in Compound doing similar things as this PR! (last time I looked was the old interface)

Reply:
Yes, but yours is better because of

> allows the record field selectors to be used directly

Reply:
I tried to support that by doing 

```.hs
data FooHandle e = { foo :: forall es. e :> es => Bar -> Eff es Baz }
```

but it wasn't ergonomic to define handlers that way. Your version is to use the simple handle definition

```
data FooHandle e = { foo :: Bar -> Eff e Baz }
```

and then put `forall es. e :> es =>` on the outside. It works out much nicer!